### ISSUE TITLE
Integration: Caldera Forms

### ISSUE THREAD

**Issue by michaelbeil** [AffiliateWP/AffiliateWP#1787]
It might be a good idea to add an integration for Caldera Forms. 

Ref: https://secure.helpscout.net/conversation/272708245/45996/?folderId=205524

**Comment by sumobi** [comment]
Do we still want to do this for 2.0? If so, we'll need to get some of their payment add-ons (probably just Stripe): https://calderaforms.com/caldera-forms-add-ons/#/payment

**Comment by pippinsplugins** [comment]
@sumobi could you request copies of them? If we can get it done, awesome, otherwise we can punt.

**Comment by sumobi** [comment]
I've reached out to them

**Comment by sumobi** [comment]
Added PR - WIP

**Comment by sumobi** [comment]
Yeahhh tag team

**Comment by pippinsplugins** [comment]
Pinging @Shelob9. The notes above may interest you :)

**Comment by pippinsplugins** [comment]
@sumobi can you please put the add-ons in Dropbox?

**Comment by Shelob9** [comment]
> Setting referrals to "unpaid" on a successful payment
> 
> There's no generic "payment completed" hook (like Gravity Forms)
> There are 4 payment gateways ("processors") for Caldera Forms. Of those, only Stripe and Braintree have payment completed hooks. PayPal and Authorize.net do not. For now we'll have 2 write 2 methods to mark the referral as unpaid.

This is correct. We do not have [this concept](https://github.com/CalderaWP/Caldera-Forms/issues/1076). The [caldera_forms_submit_complete](https://calderaforms.com/doc/caldera_forms_submit_complete/) hook will run when a submission has passed all validation and is about to be saved. 

For Paypal, in the Callback on that hook, do this:
```
    global $transdata;
    if( ! empty( $transdata['paypal_express'] ) && ! empty( $transdata['paypal_express']['result'] ) ){
        //$transdata['paypal_express']['result'] is info returned from PayPal about success
    
    }elseif( ! empty( $transdata['note'] ) ){
        //$transdata['note'] is error from paypal
    }else{
        //Things went really wrong
    }
```

For authorize.net there really is no good way to do this. @shelob9 hangs head in shame.

@sumobi What is your deadline on this? I can add hooks to each of them to make your life easy -- but that would mean your integration would require the new version of the add-ons. If you want to do that, please let me know what data you need in the callback.

> Linking the reference (entry ID) to the standalone entry page.
> 
> There's no direct entry page so we can't link the reference (entry ID) from the referrals screen. Entries are opened in modal windows.
This is not a thing we have yet. It will be once the new entry viewer is done, but that's probably 2 versions out. @sumobi I suggest a separate issue to add that link later. But make sure you are tracking entry ID.


> Rejecting referrals on refund
> 

There is no refund hook that I can see, or a hook when an entry is deleted.
 We do not process refunds through Caldera Forms. 

>Listing products in the referral
>
>I haven't found a way to get a list of products for the entry since there are no product fields. >Essentially any field can be used for the final calculation. Currently it just lists the form name.
There is no concept of a product. Our EDD add-on adds an auto-population field for EDD products, but that's it. Sometimes people do auto-populated field based on Woo's product post type, but typically they are hardcoding their product/ services to a select field settings.

>Testing Braintree
>
>I haven't been able to successfully test Braintree, it's showing an error before I submit the form of >"Actual Plan is required". I have One Time Payment selected in the processor options. It seems this >error is related to the Subscription Plan option so I suspect this might be a bug.
@sumobi Can you tell me which version of Braintree add-on and Caldera Forms you are using? Your error is confusing to me as single payment and payment plan  are two totally different options. Do you have both processors?


> Testing Stripe
> 
> Make sure the form has ajax submissions enabled. When turned off, the Stripe payment window opens and then closes after a second, then shows a No Stripe token found, please wait for the page to load fully and try again. message. Tested without any other plugins active and this is still the case.

Interesting bug, I can confirm this and will get it fixed.

**Comment by Shelob9** [comment]
@sumobi I left notes on the PR. My major concern is that you are assuming that the first calculation field added to the form represents the total passed to processor. This is a major problem because:
1) There could be 2 or more calculation fields.
2) There may be no calculation fields. Total charge could come from a different field type value or be hardcoded to the processor settings.

What would solve this problem and solve some of the problems above is to create an AffiliateWP processor with one setting: total. That way the end user can explicitly set how much the conversion is worth.

Here is a good start for that processor that you can plug your tracking into:

https://gist.github.com/Shelob9/560838531d4975e3e11850ae5e6c4d11

**Comment by sumobi** [comment]
Thanks @Shelob9,

> @sumobi What is your deadline on this? I can add hooks to each of them to make your life easy -- but that would mean your integration would require the new version of the add-ons. If you want to do that, please let me know what data you need in the callback.

We're hoping for beta around the 11th of January, and then 2 weeks later we'll release 2.0. If you can add the hooks then that would be great. All we'd need (as far as I can see) is the entry ID.

> We do not process refunds through Caldera Forms.

That's fine. Could we get a hooks for when an entry is trashed, and when it is restored? We do this in Ninja Forms to revoke the referral if the entry is trashed, and restore the referral when the entry is restored.

> I left notes on the PR. My major concern is that you are assuming that the first calculation field added to the form represents the total passed to processor.

Great point and idea! I'll look into adding an AffiliateWP specific total field.

Cheers

**Comment by sumobi** [comment]
@Shelob9 Thanks for the gist, very helpful.

@AffiliateWP/core There's now a processor that must be added in order to control the total. This can map an existing field (calculation field or something else) or you can define a total to base the referral amount off.

It's important to note that the AffiliateWP processor must come first in the list (you can drag and drop). This is so the referral is correctly marked as `unpaid` on a successful payment by the payment processor.

![screen shot 2016-12-30 at 8 18 21 pm](https://cloud.githubusercontent.com/assets/52581/21561126/65e4db28-cecd-11e6-81c3-b7844014d2ab.png)

![screen shot 2016-12-30 at 8 18 27 pm](https://cloud.githubusercontent.com/assets/52581/21561127/6ce64d62-cecd-11e6-841d-e12d4049443c.png)

**Comment by sumobi** [comment]
@AffiliateWP/core Ready for testing. Add-ons in Dropbox. 

#### Currently supports:

1. Stripe
2. PayPal
3. Braintree (untested until bug resolved).

#### Things to test:

1. Product forms with affwp processor but no payment processor. Referrals should stay at `pending` status.

2. Product forms with both affwp processor and payment processor (make sure affwp processor is first in order). Referrals should move to `unpaid` status on successful payment.

3. General lead tracking forms (no affwp processor set up because no total field). Referral should move to `unpaid` status and have a referral total of `$0.00`.

**Comment by Shelob9** [comment]
@sumobi Can I ask why this has to be tied to a payment processor at all? This introduces and arbitrary rule to the system that limits it usefulness. If you let the user determine this rule, it would be way more useful. Your way of doing this is way more rigid then it needs to be and makes assumptions that don't match how I see Caldera Forms being used.

If you removed that requirement, a user might choose to require a successful payment. That would be easy. They would simply place a payment processor in the form, if it failed, the conversion would not be recorded. If you wanted, you could in your pre_processor callback  set the conversion to pending, then change its status once the payment is run, just like I suggested.

Also, this would allow your integration to work with other payment processors then the ones you have supported or we have created.

**Comment by sumobi** [comment]
@Shelob9 

1. Can you show me how to retrieve the entry ID from within the `pre_processor` method, and in other places, if given the `$form` object ? I saw you left a note about not using `get_submission_data()`.

2. The reason I have the additional methods for each payment processor is because the `processor()` method inside the `AffiliateWP_Caldera_Forms_Processor` class runs even if there's no payment processor added to the form. 

    If I set the referral to be `unpaid` within this method, the site admin will lose money to affiliates for payments they did not receive. Granted, the site admin should always add a payment processor for any form with a total field, but I was trying to minimise the losses :) Am I missing something here? Should that method not run at all unless there's a payment processor added?

I'd love to remove the individual methods for each payment processor if I can. Ideally I'd set the referral to `pending` in `pre_processor()` (having trouble getting the entry ID there) like you suggested and then complete the referral in the `processor()` method. But this should only happen if a successful payment is made, otherwise the referral stays `pending`.

**Comment by sumobi** [comment]
I 100% agree with you Josh, I'm still wrapping my head around how Caldera Forms works with processors and what not. I've updated the PR so now it's independent of the payment processor.

**Comment by sumobi** [comment]
@AffiliateWP/core Ready for some more testing

**Comment by pippinsplugins** [comment]
@sumobi works great! Had no issues at all with it.

Could we update `Total` here to something more meaningful?
![screen shot 2017-01-03 at 9 15 07 pm](https://cloud.githubusercontent.com/assets/1034109/21630744/be631f10-d1f9-11e6-9ca0-72ce7c65b018.png)

That screen is not super clear that the `Total` represents the calculation base for referrals. Perhaps just some help text like `This total will be used to calculate the referral amount. For example, if the referral rate is 10% and the total here is $100, the referral will be created at $10.`

**Comment by sumobi** [comment]
👍

**Comment by sumobi** [comment]
@pippinsplugins added

![screen shot 2017-01-04 at 5 08 52 pm](https://cloud.githubusercontent.com/assets/52581/21631478/808f4770-d2a0-11e6-8642-05c3abfdf906.png)

**Comment by Shelob9** [comment]
@sumobi I'm glad that you made it independent of the processor. Please let me know when this launches. We will do a blog post about it and add a page to our site about it with a link back to your site.

**Comment by pippinsplugins** [comment]
@sumobi Perfect. I'm happy with this.

**Comment by sumobi** [comment]
Sounds excellent @Shelob9, we'll ping you when v2.0 is out

**Comment by pippinsplugins** [comment]
Merging and closing.
